### PR TITLE
Update linear_search.py

### PR DESCRIPTION
**Pull request by Ashley-J-George**
I've added a Python implementation of recursive search algorithm

### **Describe your change:**

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### **Checklist:**
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [ ] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [ ] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

**Comment by spamegg1** [comment]
@Ashley-J-George Hi there,
It looks like you did not provide any doctests for `rec_linear_search`.
`rec_linear_search` still has the same doctests as `linear_search`.
Can you provide some different doctests that use `rec_linear_search` instead?

There is an issue when, for example, `low` or `high` are greater than `len(sequence)`. It would give a list index out of bounds error. Maybe handle it with an exception?

Also, type hints for both `linear_search` and `rec_linear_search` would be appreciated. For example:
```python
def rec_linear_search(sequence: list, low: int, high: int, target: float) -> int:
```

File: searches/linear_search.py [hunk 1]
```diff
@@ -38,6 +38,39 @@ def linear_search(sequence, target):
     return None
 
 
+def rec_linear_search(sequence, low, high, target) -> int:
```
Comment (cclauss) [inline searches/linear_search.py:41]: Type hints on function parameters?

File: searches/linear_search.py [hunk 2]
```diff
@@ -38,6 +38,39 @@ def linear_search(sequence, target):
...
+    4
+
+    >>> linear_search([0, 5, 7, 10, 15], 5)
+    1
+
+    >>> linear_search([0, 5, 7, 10, 15], 6)
+
+    '''
+    if high < low:
+        return None
```
Comment (cclauss) [inline searches/linear_search.py:66]: Should this raise a ValueError for invalid input?

Comment (Ashley-J-George) [reply searches/linear_search.py:66]: This is the condition that tells us that high has crossed low and the element is not in the list

File: searches/linear_search.py (hunk 1 above)
Comment (Ashley-J-George) [reply searches/linear_search.py:41]: I'll add it

File: searches/linear_search.py (hunk 2 above)
Comment (cclauss) [reply searches/linear_search.py]: We promised on line 41 that this function would return an it but now we are returning an non-int.  We have two choices:
1. Return -1 as a signal that the element is not in the list --or--
2. Raise an error.

This is like [str.find()](https://docs.python.org/3/library/stdtypes.html#str.find) vs. [str.index](https://docs.python.org/3/library/stdtypes.html#str.index).

**Comment by Ashley-J-George** [comment]
Hey @Ashley-J-George,
Something went wrong with the build.

TravisCI finished with status errored, which means the build failed because of something unrelated to the tests, such as a problem with a dependency or the build process itself.


@cclauss @spamegg1  How should I rectify this?

**Comment by Ashley-J-George** [comment]
When the element is not found I have made the function return -1, and added
a try except to handle the index error

**Comment by spamegg1** [comment]
@Ashley-J-George Build log says
> 3     W293 blank line contains whitespace

Very easy to fix! Check out [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md) about running `black` on your code.

File: searches/linear_search.py [hunk 3]
```diff
@@ -38,6 +38,41 @@ def linear_search(sequence, target):
     return None
```
Comment (cclauss) [inline searches/linear_search.py:38]: Line 14 promises that we are going to return an int but line 38 returns None.  We should not surprise the caller.

File: searches/linear_search.py [hunk 4]
```diff
@@ -38,6 +38,41 @@ def linear_search(sequence, target):
...
+
+    :param sequence: An array of items
+    :param low: Lower bound of the array
+    :param high: Higher bound of the array
+    :param target: The element to be found
+    :return: Index of the key or -1 if key not found or None in case of an exception
+
+    Examples:
+    >>> rec_linear_search([0, 30, 500, 100, 700], 0, 4, 0)
+    The element 0 is present at index 0
```
Comment (cclauss) [inline searches/linear_search.py:53]: This function returns an int.
```suggestion
    0
```

File: searches/linear_search.py [hunk 5]
```diff
@@ -38,6 +38,41 @@ def linear_search(sequence, target):
...
+    if high < low:
+        return -1
+    try:
+        if sequence[low] == target:
+            return low
+        if sequence[high] == target:
+            return high
+        return rec_linear_search(sequence, low + 1, high - 1, target)
+    except IndexError:
+        print('Invalid upper or lower bound!')
```
Comment (cclauss) [inline searches/linear_search.py:73]: CONTRIBUTING.md says that `algorithmic functions should not print()` so raise Error.

**Review by cclauss — APPROVED** [review]
Nice work!!

**Comment by Ashley-J-George** [comment]
Thank You!

**Pull request merged**